### PR TITLE
Handle WM_MEASUREITEM message to set owner-drawn combo boxes height properly.

### PR DESCRIPTION
**Pull request by beru**
Hi, this PR fixes the problem that owner-drawn combo boxes are not fully displayed on High DPI screens.

This is the example picture.
ss

### CODE REVIEW

File: src/TortoiseMerge/SetMainPage.cpp [hunk 1]
```diff
@@ -266,3 +267,17 @@ BOOL CSetMainPage::DialogEnableWindow(UINT nID, BOOL bEnable)
 	}
 	return pwndDlgItem->EnableWindow(bEnable);
 }
+
+void CSetMainPage::OnMeasureItem(int nIDCtl,LPMEASUREITEMSTRUCT lpMeasureItemStruct)
+{
+	CFont* pFont = this->GetFont();
+	if (pFont != NULL) {
```
Comment (csware) [inline src/TortoiseMerge/SetMainPage.cpp]: please don't use `NULL`, but `!pFont` (preferred) or `pFont != nullptr`

File: src/TortoiseMerge/SetMainPage.cpp [hunk 2]
```diff
@@ -266,3 +267,17 @@ BOOL CSetMainPage::DialogEnableWindow(UINT nID, BOOL bEnable)
...
 }
+
+void CSetMainPage::OnMeasureItem(int nIDCtl,LPMEASUREITEMSTRUCT lpMeasureItemStruct)
+{
+	CFont* pFont = this->GetFont();
+	if (pFont != NULL) {
+		CDC* pDC = GetDC();
+		CFont* pFontPrev = pDC->SelectObject(pFont);
+		int iborder = ::GetSystemMetrics(SM_CYBORDER);
+		CSize sz = pDC->GetTextExtent(_T("0"));
```
Comment (csware) [inline src/TortoiseMerge/SetMainPage.cpp]: Please use `L"0"`

File: src/TortoiseMerge/SetMainPage.h [hunk 3]
```diff
@@ -53,6 +53,7 @@ class CSetMainPage : public CPropertyPage
 	afx_msg void OnModified();
 	afx_msg void OnModifiedWithReload();
 	afx_msg void OnBnClickedWhitespace();
+    afx_msg void OnMeasureItem(int nIDCtl,LPMEASUREITEMSTRUCT lpMeasureItemStruct);
```
Comment (csware) [inline src/TortoiseMerge/SetMainPage.h]: Please make sure indention and whitepsace are correct.

File: src/TortoiseMerge/SetMainPage.cpp [hunk 4]
```diff
@@ -266,3 +267,17 @@ BOOL CSetMainPage::DialogEnableWindow(UINT nID, BOOL bEnable)
 	}
 	return pwndDlgItem->EnableWindow(bEnable);
 }
+
+void CSetMainPage::OnMeasureItem(int nIDCtl,LPMEASUREITEMSTRUCT lpMeasureItemStruct)
```
Comment (csware) [inline src/TortoiseMerge/SetMainPage.cpp]: Please add whitespace after the `,`

File: src/TortoiseMerge/SetMainPage.cpp [hunk 5]
```diff
@@ -266,3 +267,17 @@ BOOL CSetMainPage::DialogEnableWindow(UINT nID, BOOL bEnable)
...
+	CFont* pFont = this->GetFont();
+	if (pFont != NULL) {
+		CDC* pDC = GetDC();
+		CFont* pFontPrev = pDC->SelectObject(pFont);
+		int iborder = ::GetSystemMetrics(SM_CYBORDER);
+		CSize sz = pDC->GetTextExtent(_T("0"));
+		lpMeasureItemStruct->itemHeight = sz.cy + 2 * iborder;
+		pDC->SelectObject(pFontPrev);
+	}
+    CPropertyPage::OnMeasureItem(nIDCtl,lpMeasureItemStruct);
```
Comment (csware) [inline src/TortoiseMerge/SetMainPage.cpp]: Whitesapce after the `,`

**Comment by csware** [comment]
Please squash the commits.

**Comment by beru** [comment]
> It would be great if there is also a comment in the commit message why this is needed. We use CMFCFontComboBox so I'd expect that this is already properly handled.

I will add more explanation to commit messages next time. And sorry about coding style problems. I fixed them according to your comments.

I had expected `CMFCFontComboBox` to handle the matter but it doesn't. I wish Microsoft would put more efforts into legacy platform.

Please look at `CMFCFontComboBox::MeasureItem` in below file.
`C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\atlmfc\src\mfc\afxfontcombobox.cpp`
And compare the code with `CMyComboBox::MeasureItem` written in
https://msdn.microsoft.com/library/4e73b5df-0d2e-4658-9706-38133fb10513.aspx#Anchor_39

`CMFCFontComboBox::MeasureItem` implementation is no good.

**Comment by csware** [comment]
Wouldn't it be better to sub-class CMFCFontCombobox and fix the implementation of `MeasureItem`?

**Comment by beru** [comment]
What would be the name of the extended class?
Here are some candidates...
- CTGitFontCombobox
- CMFCFontComboboxEx

**Comment by csware** [comment]
I don't care for the name. However, I'd try to avoid just adding "Ex" because M$ could also do this...

**Comment by beru** [comment]
@csware I've tried to implement the extended class that handles the matter, but failed in vein.

I've found out `CMFCFontComboBox::MeasureItem` isn't for control height but for its list items heights.
https://msdn.microsoft.com/en-us/library/12h9x0ch.aspx#ccombobox__measureitem
> If the combo box is created with the CBS_OWNERDRAWVARIABLE style, the framework calls this member function for each item in the list box. Otherwise, this member is called only once.

Add to that, It seems like handling `WM_MEASUREITEM` in control itself is not possible.
https://msdn.microsoft.com/ja-jp/aa249909
> If you subclass by using DDX_Control, SubclassDlgItem, or SubclassWindow, the subclassing usually occurs after the creation process. Therefore, there is no way to handle the WM_MEASUREITEM message in the control's OnChildNotify function, which is the mechanism MFC uses to implement ON_WM_MEASUREITEM_REFLECT.

Handling `WM_MEASUREITEM` message in parent window may be not elegant but suffice the goal.

By the way, I forgot to add `ReleaseDC` calls, so pushed another commit.

**Comment by csware** [comment]
Merged with commit 9575d058ebdeb02adbbe6f05a9193acb10d9966f.

**Comment by csware** [comment]
Thanks for bringing this up and also providing a patch!

**Pull request closed without merging**